src/test/java/uk/gov/hmcts/reform/sscs/service/exceptions/FeignClientErrorDecoderTest.java

```diff
@@ -0,0 +1,164 @@
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import feign.Request;
+import feign.RequestTemplate;
+import feign.Response;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.server.ResponseStatusException;
+import uk.gov.hmcts.reform.sscs.model.Message;
+import uk.gov.hmcts.reform.sscs.model.single.hearing.CaseDetails;
+import uk.gov.hmcts.reform.sscs.model.single.hearing.HearingRequestPayload;
+import uk.gov.hmcts.reform.sscs.service.AppInsightsService;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@ExtendWith(MockitoExtension.class)
+class FeignClientErrorDecoderTest {
+
+    private static final String ID = "id";
+    private static final Long CASE_ID = 1000000000L;
+    private static final String ERROR_MSG = " \"Error in calling the client method:someMethod\"";
+
+    private FeignClientErrorDecoder feignClientErrorDecoder;
+    private HearingRequestPayload hearingRequestPayload;
+
+    @Mock
+    private AppInsightsService appInsightsService;
+
+    @BeforeEach
+    void setUp() {
+        feignClientErrorDecoder = new FeignClientErrorDecoder(appInsightsService);
+        hearingRequestPayload = new HearingRequestPayload();
+        hearingRequestPayload.setCaseDetails(new CaseDetails());
+        hearingRequestPayload.getCaseDetails().setCaseRef(String.valueOf(CASE_ID));
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {400, 401, 403, 404})
+    void should_handle_4xx_post_put_error(int statusCode) throws JsonProcessingException {
+        Map<String, Collection<String>> headers = new HashMap<>();
+        Request request =
+            Request.create(Request.HttpMethod.POST, "url",
+                           headers, Request.Body.create(toJsonString(hearingRequestPayload)), null);
+
+        Response response = Response.builder()
+            .request(request)
+            .status(statusCode)
+            .reason("Error")
+            .body("Some body", StandardCharsets.UTF_8)
+            .build();
+
+        Throwable throwable = feignClientErrorDecoder.decode("someMethod", response);
+        verify(appInsightsService, times(1)).sendAppInsightsEvent(any(Message.class));
+        assertThat(throwable).isInstanceOf(ResponseStatusException.class);
+
+
+        if (statusCode == 400) {
```

Comment:
Ideally this should be split out into different tests. You can pass in the expected results as a Valuesource. Adding logic into the tests adds another point of failure.

Reply:
I don't necessarily agree with this but in the interest of time it's fine